synapse/storage/databases/state/bg_updates.py

```diff
@@ -98,81 +98,111 @@ def _get_state_groups_from_groups_txn(
-                        where_clause = "(type = ?)"
-                        overall_select_query_args.extend([etype])
-                    else:
-                        where_clause = "(type = ? AND state_key = ?)"
-                        overall_select_query_args.extend([etype, skey])
-
-                    select_clause_list.append(
+        # This is an optimization to create a select clause per-condition. This
+        # makes the query planner a lot smarter on what rows should pull out in the
+        # first place and we end up with something that takes 10x less time to get a
+        # result.
+        use_condition_optimization = (
+            not state_filter.include_others and not state_filter.is_full()
+        )
+        state_filter_condition_combos: List[Tuple[str, Optional[str]]] = []
+        # We only need to caclculate this list if we're using the condition optimization
+        if use_condition_optimization:
+            for etype, state_keys in state_filter.types.items():
+                if state_keys is None:
+                    state_filter_condition_combos.append((etype, None))
+                else:
+                    for state_key in state_keys:
+                        state_filter_condition_combos.append((etype, state_key))
+        # And here is the optimization itself. We don't want to do the optimization
+        # if there are too many individual conditions. 10 is an arbitrary number
+        # with no testing behind it but we do know that we specifically made this
+        # optimization for when we grab the necessary state out for
+        # `filter_events_for_client` which just uses 2 conditions
+        # (`EventTypes.RoomHistoryVisibility` and `EventTypes.Member`).
+        if use_condition_optimization and len(state_filter_condition_combos) < 10:
+            select_clause_list: List[str] = []
+            for etype, skey in state_filter_condition_combos:
+                if skey is None:
+                    where_clause = "(type = ?)"
+                    overall_select_query_args.extend([etype])
+                else:
+                    where_clause = "(type = ? AND state_key = ?)"
+                    overall_select_query_args.extend([etype, skey])
+
+                # Small helper function to wrap the union clause in parenthesis if we're
+                # using postgres. This is because SQLite doesn't allow `LIMIT`/`ORDER`
+                # clauses in the union subquery but postgres does as long as they are
+                # wrapped in parenthesis which this function handles the complexity of
+                # handling.
+                def wrap_union_if_postgres(
+                    union_clause: str, extra_order_or_limit_clause: str = ""
+                ) -> str:
+                    if isinstance(self.database_engine, PostgresEngine):
+                        return f"""({union_clause} {extra_order_or_limit_clause})"""
+
+                    return union_clause
```

Comment:
It would be better just to define this function somewhere above once but thought it lost the specific usage context when we put it somewhere else. Feel free to poke :fast_forward: